### ISSUE TITLE
[Feature Request]: Linux/PulseAudio - Allow changing the PRODUCT_STRING of the embedded Chromium

### ISSUE THREAD

**Issue by Luzifer** [electron/electron#27581]
**Preflight Checklist**

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for a feature request that matches the one I want to file, without success.

**Problem Description**

Running multiple Electron applications yields multiple `Chromium: Playback` entries in PulseAudio.

As a user having the requirement to have certain audio source mapped to specific sinks this causes an issue for me: I cannot determine which of those is my browser and which belongs to which Electron application.

Therefore I cannot move a specific audio stream to the target sink. And if I can figure out which one to move PulseAudio remembers "Chromium" to be moved to that sink.

**Proposed Solution**

Chromium contains a constant which is set only once and is used for PulseAudio identifier. If this constant would be set to the application name the applications could be distinguished easily and the assignment to a specific sink should work properly.

Code to be patched:
https://github.com/chromium/chromium/blob/99314be8152e688bafbbf9a615536bdbb289ea87/media/audio/pulse/pulse_util.cc#L39-L42

**Alternatives Considered**

Compiling an own release of Electron every time I need an app built but that would be a massive overkill. Having a dynamic way to use the AppName would be a way better solution…

**Additional Information**

_No response_

**Comment by Zhincore** [comment]
No news on this? :(

**Comment by tauin** [comment]
should be noted that it seems to cause [this](https://github.com/vector-im/element-web/issues/10928) bug in element (matrix client)

**Comment by awfulcooking** [comment]
Affects Discord, too.

**Comment by starquake** [comment]
I think this also makes the KDE feature, that shows which application is generating sound, show on Chromium

**Comment by Vinjul1704** [comment]
In addition to setting it per-app, it would be nice if you could change it for each individual audio stream. Certain applications, such as chat/voice programs like Discord (or Webcord in my particular case), can create multiple input streams in order to capture both microphone and desktop audio at the same time (talking + desktop audio sharing, for example). The same applies for output streams of other types of apps too, of course, like one for notification/UI audio and one for music.

In this case, having both use the same name, even if it's different from other electron apps, can cause issues like selected devices being mixed up, system volume controls being saved incorrectly and in general overall confusion.

**Comment by Sqwrly** [comment]
Any updates? It's pretty maddening opening the volume control and seeing 5 different instances of Chromium and they are all Electron apps. This prevents routing apps to different sinks. 😔

**Comment by TheInfamousAlk** [comment]
Any updates on this? I'm still struggling daily with this. 

Linking this as well since it doesn't seem to be linked, bug report about this issue that was closed for some reason: https://github.com/electron/electron/issues/27581

**Comment by lectrode** [comment]
This looks like the original thread where `AudioManager::GetGlobalAppName` was created. However, it doesn't look like the implementation was ever finished chromium-side (at least from that conversation).

https://groups.google.com/a/chromium.org/g/media-dev/c/v8JtLAvZiIQ

**Comment by KaKi87** [comment]
Hi,

I tried bumping the linked thread, but apparently that place is read-only for normal people.

Could someone with access bump it please ?

Thanks

**Comment by roliverio** [comment]
Struggling as well to have Electron apps have personalized names for PA or Pipewire. Thanks for any effort implementing this.

**Comment by seythegamer** [comment]
Weird but GNOME (Fedora) makes all the electron-chormium apps have their own corresponding names, but it just doesnt work in KDE Plasma.

<img width="365" height="401" alt="Image" src="https://github.com/user-attachments/assets/d475ca3a-558e-4aaa-adce-1ba792c83723" />

**Comment by ZeroKnight** [comment]
From a user perspective, it's pretty disappointing that in a few months it'll be the 5th anniversary of this issue and there's been no movement on it yet. Especially considering it affects every Electron app that makes use of audio on Linux.

**Comment by sosloow** [comment]
> Weird but GNOME (Fedora) makes all the electron-chormium apps have their own corresponding names, but it just doesnt work in KDE Plasma.

@seythegamer wait, for me it doesn't, in settings -> audio -> volume levels, at least. It's all Chromium. What's on your screenshot? Is this an extension, or some widget in the default shell you need to enable?

**Comment by josselinonduty** [comment]
> From a user perspective, it's pretty disappointing that in a few months it'll be the 5th anniversary of this issue and there's been no movement on it yet. Especially considering it affects every Electron app that makes use of audio on Linux.

I completely agree. This bug that has been causing problems with [deezer-linux](https://github.com/aunetx/deezer-linux) for years is not being fixed by either Chromium or Electron.

Given Linux’s growing market share on desktop (and even mobile), it is very disappointing.

**Comment by Damglador** [comment]
I love that Google made an exception that changes the name to Google Chrome, but didn't think at all about making the thing not hard codded.

**Comment by Damglador** [comment]
Does anyone know how to get proper icon name for the app running? The PRODUCT_STRING (rn Chromium) is used for the stream name, and it doesn't matter what it gets, it'll be better than whatever it is now anyway, but kBrowserDisplayName (rn chromium-browser) has to be a valid xdg icon name, so it needs the name of the icon of the current app. Leaving it as is and changing the text would still be better than nothing, but it would be nice to have proper icons as well.

Relevant doc: https://www.freedesktop.org/wiki/Software/PulseAudio/Documentation/Developer/Clients/ApplicationProperties/#pa_prop_application_icon_name
Relevant code: https://github.com/chromium/chromium/blob/5f49b002c76dab410e718ae3b01497a1ce27dc22/media/audio/pulse/pulse_util.cc#L574-L578

**Comment by Juts2** [comment]
Can we please get some traction or a response on this? Its crazy

**Comment by GeorgeBroughton** [comment]
I have the same problem in CachyOS/Wayland/Hyprland/Quickshell

One of these is Discord (Electron), the other an SSB for YouTube Music

<img width="404" height="646" alt="Image" src="https://github.com/user-attachments/assets/632129c7-b929-44c3-92a9-fa3359277825" />

**Comment by lipidconcentrategaming** [comment]
Same problem on CachyOS/KDE 

<img width="424" height="295" alt="Image" src="https://github.com/user-attachments/assets/d88076e7-0d7f-4cd4-8acb-6dca2eed3704" />

**Comment by Damglador** [comment]
I wonder how Vencord got an icon

**Comment by Kesefon** [comment]
I played around with the code a bit and noticed that the current method of setting the stream name is broken because all the audio stuff got moved into a separate process.

Which means that for a proper fix, `AudioManager::SetGlobalAppName()` needs to be called using some sort of ipc mechanism, that I am not smart enough for.

BUT as a workaround, you can run your electron app with `--disable-features=AudioServiceOutOfProcess` which makes the old code work again.



This worked for me with VSCode, Signal Desktop, and Cookie Clicker (electron version 39 & 32).

<img width="462" height="420" alt="Screenshot of volume applet with working names" src="https://github.com/user-attachments/assets/b3954308-abc9-4cad-867b-e83dbd60626f" />

If you're wondering why only Signal has an icon, that seems to be based on the name of the executable; i.e. rename your `electron` binary to `signal-desktop` and you get a signal icon. This is a fallback handled by Plasma and limited to system-wide installed icons (I think; YMMV).

**Comment by Damglador** [comment]
That explains why all my attempts to pass a string didn't work and why pulse_util.cc doesn't log anything.

**Comment by Damglador** [comment]
`--disable-features=AudioServiceOutOfProcess` fixes the problem partially.

For some reason, `CreateOutputStream()` creates a pulseaudio context in itself, while `CreateInputStream()` receives context from arguments. The received context is created in `InitPulse()`, which right now just uses the hard-codded `PRODUCT_STRING` + " input". Ignoring the fact that " input" is completely redundant there, using `AudioManager::GetGlobalAppName()` in there seems to not be possible, likely because it is initialized before `GlobalAppName` is set (so it receives null).

I've found a shitty workaround for this - just using an empty string. This creates a context with no name, but makes Plasma's audio manager display `media.name` instead, which can be easily set in `CreateInputStream()` by passing `AudioManager::GetGlobalAppName()` to `pa_stream_new_with_proplist()`, which right now recieves hard-codded "RecordStream". I would consider this better than having a bunch of "Chromium input"

Also, for the icon I used argv0, just get it and pray there's an icon that matches the name. It should work most of the time, since usually binary name matches the icon name, but if it doesn't - not a big deal. Telegram for example doesn't assign an icon at all. Sadly, it does mean that packages that use system Electron will get the Electron icon (or no icon) instead of their own one.

In my opinion, that's better than whatever it is now, but I doubt if it's good enough for a PR. So I'll just leave the code here: https://github.com/electron/electron/compare/main...Damglador:electron:main

<img width="442" height="412" alt="Image" src="https://github.com/user-attachments/assets/068d311f-385b-4a69-9e15-65881315f82c" />

**Comment by Kesefon** [comment]
I only did very limited testing on this, but `platform_util::GetXdgAppId()` seems to work even when called from the out of process audio service. So I made a quick patch that uses this for the icon. And for the name it calls `electron::GetPossiblyOverriddenApplicationName()` which in turn uses the XdgAppId to extract the application name from a .desktop file. 

see here: https://github.com/Kesefon/electron/commit/5f798130d374812e7a9cbc86215546861bd6f9b9

For the one app I've tested, everything worked ootb and it even resulted in a better name being displayed since `app.name` was just some technical id and not a proper name.

The obvious drawbacks are that it only works in combination with a .desktop file and later name changes are still broken.


----------

Some quick thoughts regarding the different behaviour of `CreateInputStream()` and `CreateOutputStream()`:
- I wonder why it's implemented this way?
  - What would happen if we make both work the same way?
  - What does the `context` even do?
    - Can we create infinite contexts or are you supposed to reuse them application-wide?
- Looks like there is a [`PROP_APPLICATION_NAME`](https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/blob/master/vala/libpulse.vapi#L580)
  - Could we set that the same way as `PROP_APPLICATION_ICON_NAME`?
- Also cool to see that you already have a workaround 

I will investigate this some more later

**Comment by Damglador** [comment]
From my understanding, context is supposed to be application-wide. With pipewire-pulse, context name sets the `node.name` and `application.name` (shown in `pw-dump`). Using the `node.name` you can dump info about a particular app, for example: `pw-dump Chromium`.

On a side note, can electron be compiled with `--disable-features=AudioServiceOutOfProcess` by default? I don't know what out of process audio is supposed to solve, but it clearly introduces a lot of issues.

**Comment by Kesefon** [comment]
There is a built in disable features list https://github.com/electron/electron/blob/809ab09b6f5ede0a9f9b0fcc0e9c33ff971277f2/shell/browser/feature_list.cc#L47-L65 
You can add `features::kAudioServiceOutOfProcess.name` here.

After skimming the design docs [[1]](https://docs.google.com/document/d/1s_Fd1WRDdpb5n6C2MSJjeC3fis6hULZwfKMeDd4K5tI/edit?tab=t.0) [[2]](https://docs.google.com/document/d/1fVHVJjd5zTqc6O7b0lprILyuFPC9qVQUudzjEgOq1HY/edit?tab=t.0#heading=h.ydr456a07ivp) it seems the main purpose of out of process audio is reliability/crash resistance with some performance and security benefits aswell.

It seems the intention is to remove in-process audio support long term (https://crbug.com/40066532). Not sure how actively this is pursued though.

**Comment by Damglador** [comment]
I'm sure it's gonna be fine without it. Hopefully they'll fix the naming problem upstream before removing the flag.
